### PR TITLE
[RFXCOM] Changes for timeouts in TcpConnector

### LINKED ISSUE #2432: Changes in #2101 seem not compatible with TcpConnector
See also https://community.openhab.org/t/rfxcom-error-occurred-during-packet-receiving/30939/5

### PR DESCRIPTION
**Pull request by martinvw**
The tcp-stream behaves different when a timeout is set and terminates with an exception instead this should be properly handled. 

Fixes #2432

Original changes where done in https://github.com/openhab/openhab2-addons/pull/2101

### CODE REVIEW

**Comment by martinvw** [comment]
@mjagdis can you please take a look whether this makes sense, I personally do not yet like the SocketTimeoutException bleeding in this class, so I am open for suggestions.

**Comment by martinvw** [comment]
Download: [org.openhab.binding.rfxcom-2.2.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab2-addons/files/1122991/org.openhab.binding.rfxcom-2.2.0-SNAPSHOT.jar.zip)

**Comment by mjagdis** [comment]
Why not change RFXComStreamReader to use connector.read. Then the connector classes just return in.read except for the socket connector which also catches SocketTimeoutException and returns 0 in that case. If you do that everything is hidden in the relevant connectors, the stream reader only needs to know about the connector and you can remove the InputStream from RFXComStreamReader as well.

**Comment by martinvw** [comment]
Yes that should indeed be the solution I already considered somehow wrapping the read method as the solution this night 😉 i was actually considering to return a wrapped inputstream, but I ll try to take a look at it tonight.

**Comment by martinvw** [comment]
> Why not change RFXComStreamReader to use connector.read. Then the connector classes just return in.read except for the socket connector which also catches SocketTimeoutException and returns 0 in that case. If you do that everything is hidden in the relevant connectors, the stream reader only needs to know about the connector and you can remove the InputStream from RFXComStreamReader as well.

I checked the source now (decided not to do that while I was already in bed) indeed, that makes complete sense.

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComBaseConnector.java [hunk 1]
```diff
@@ -22,7 +24,8 @@
 public abstract class RFXComBaseConnector implements RFXComConnectorInterface {
     private final Logger logger = LoggerFactory.getLogger(RFXComBaseConnector.class);
 
-    private static List<RFXComEventListener> listeners = new ArrayList<RFXComEventListener>();
+    private List<RFXComEventListener> listeners = new ArrayList<>();
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComBaseConnector.java]: @Jamstah can you agree that this should not be static to properly support multiple bridges?

**Comment by martinvw** [comment]
New download:

[org.openhab.binding.rfxcom-2.2.0-SNAPSHOT.jar.zip](https://github.com/openhab/openhab2-addons/files/1125679/org.openhab.binding.rfxcom-2.2.0-SNAPSHOT.jar.zip)

File: addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComBaseConnector.java (hunk 1 above)
Comment (Jamstah) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComBaseConnector.java]: Yes, I'm not sure why that would have been static in the first place.

Comment (martinvw) [inline addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComBaseConnector.java]: @Jamstah if you have time could you do full review of this PR?

**Comment by martinvw** [comment]
FTR I just received confirmation that it works, so if we can review this it should be done

https://community.openhab.org/t/rfxcom-error-occurred-during-packet-receiving/30939/14

**Comment by martinvw** [comment]
@mjagdis if you could (also?) do a full review that would be great, thanks in advance!

**Comment by kaikreuzer** [comment]
@mjagdis I'll merge this already, but feel free to comment any time, if you find any issue - I am sure @martinvw will do follow-up PRs for all findings!

**Comment by SR-G** [comment]
Hey. Is the RFXCom binding should be able to work now with TCP devices ?

I have an old RFXLan, not configured under OpenHAB for a while, that i was willing to re-wire inside OH.

Either with OpenHAB 2.2.0-SNAPSHOT (build #1044) (from current docker image), either with the jar of the binding provided above (compiled on july 5th), i still have the same issues / behavior (see below). Is it normal (work in progress ?), or does i have to build a newer jar ?

> 2017-09-26 16:16:34.224 [INFO ] [nternal.connector.RFXComTcpConnector] - Connecting to RFXCOM at 192.168.8.13:10001 over TCP/IP
> 2017-09-26 16:16:34.224 [INFO ] [nternal.connector.RFXComTcpConnector] - Debug level is:  not set
> 2017-09-26 16:16:38.036 [ERROR] [g.rfxcom.handler.RFXComBridgeHandler] - Error occurred: Timeout during packet read

**Comment by martinvw** [comment]
Yes it should work now, so if you could help debuging it, it would be great!

**Comment by martinvw** [comment]
Could you enable debug logging?

**Comment by SR-G** [comment]
Sure, after executing `log:set DEBUG org.openhab.binding.rfxcom` in console, i then get in logs : 
`
2017-09-26 16:37:34.522 [INFO ] [nternal.connector.RFXComTcpConnector] - Debug level is:  not set
2017-09-26 16:37:37.444 [ERROR] [g.rfxcom.handler.RFXComBridgeHandler] - Error occurred: Timeout during packet read
2017-09-26 16:38:34.223 [DEBUG] [g.rfxcom.handler.RFXComBridgeHandler] - Checking RFXCOM transceiver connection, thing status = OFFLINE
2017-09-26 16:38:34.224 [DEBUG] [g.rfxcom.handler.RFXComBridgeHandler] - Connecting to RFXCOM transceiver
2017-09-26 16:38:34.224 [DEBUG] [nternal.connector.RFXComTcpConnector] - Disconnecting
2017-09-26 16:38:34.224 [DEBUG] [nternal.connector.RFXComTcpConnector] - Interrupt stream listener
2017-09-26 16:38:34.225 [DEBUG] [nternal.connector.RFXComTcpConnector] - Close tcp out stream
2017-09-26 16:38:34.227 [DEBUG] [nternal.connector.RFXComTcpConnector] - Close tcp in stream
2017-09-26 16:38:34.228 [DEBUG] [nternal.connector.RFXComTcpConnector] - Close socket
2017-09-26 16:38:34.228 [DEBUG] [nternal.connector.RFXComTcpConnector] - Closed
2017-09-26 16:38:34.228 [INFO ] [nternal.connector.RFXComTcpConnector] - Connecting to RFXCOM at 192.168.8.13:10001 over TCP/IP
2017-09-26 16:38:34.228 [INFO ] [nternal.connector.RFXComTcpConnector] - Debug level is:  debug
2017-09-26 16:38:34.229 [DEBUG] [g.rfxcom.handler.RFXComBridgeHandler] - Reset controller
2017-09-26 16:38:34.229 [DEBUG] [nternal.connector.RFXComStreamReader] - Data listener started
2017-09-26 16:38:34.530 [DEBUG] [g.rfxcom.handler.RFXComBridgeHandler] - Get status of controller
2017-09-26 16:38:48.069 [DEBUG] [nternal.connector.RFXComStreamReader] - Received exception, will report it to listeners
org.openhab.binding.rfxcom.internal.exceptions.RFXComTimeoutException: Timeout during packet read
        at org.openhab.binding.rfxcom.internal.connector.RFXComStreamReader.processMessage(RFXComStreamReader.java:81) [226:org.openhab.binding.rfxcom:2.2.0.201707051837]
        at org.openhab.binding.rfxcom.internal.connector.RFXComStreamReader.run(RFXComStreamReader.java:55) [226:org.openhab.binding.rfxcom:2.2.0.201707051837]
2017-09-26 16:38:48.070 [ERROR] [g.rfxcom.handler.RFXComBridgeHandler] - Error occurred: Timeout during packet read
2017-09-26 16:38:48.070 [DEBUG] [nternal.connector.RFXComStreamReader] - Data listener stopped
`

(with the RFXCOM jar provided above and compiled on july, 5th)

And (after purging cache/, tmp/ and restarting OH2 - so with default compiled far from 2.2.0-SNAPSHOT build #1044) : 
`
2017-09-26 16:43:15.169 [INFO ] [nternal.connector.RFXComTcpConnector] - Connecting to RFXCOM at 192.168.8.13:10001 over TCP/IP
2017-09-26 16:43:15.172 [DEBUG] [g.rfxcom.handler.RFXComBridgeHandler] - Reset controller
2017-09-26 16:43:15.172 [DEBUG] [nternal.connector.RFXComStreamReader] - Data listener started
2017-09-26 16:43:15.495 [DEBUG] [g.rfxcom.handler.RFXComBridgeHandler] - Get status of controller
2017-09-26 16:43:15.595 [DEBUG] [nternal.connector.RFXComStreamReader] - Received exception, will report it to listeners
org.openhab.binding.rfxcom.internal.exceptions.RFXComTimeoutException: Timeout during packet read
        at org.openhab.binding.rfxcom.internal.connector.RFXComStreamReader.processMessage(RFXComStreamReader.java:79) [226:org.openhab.binding.rfxcom:2.2.0.201709261356]
        at org.openhab.binding.rfxcom.internal.connector.RFXComStreamReader.run(RFXComStreamReader.java:53) [226:org.openhab.binding.rfxcom:2.2.0.201709261356]
2017-09-26 16:43:15.597 [ERROR] [g.rfxcom.handler.RFXComBridgeHandler] - Error occurred: Timeout during packet read
2017-09-26 16:43:15.598 [DEBUG] [nternal.connector.RFXComStreamReader] - Data listener stopped
`

(seems to be the same behavior)

**Comment by martinvw** [comment]
And can you telnet to that ip / port?

**Comment by SR-G** [comment]
Yep but i get just crap (which is normal).

> telnet 192.168.8.13 10001
> Trying 192.168.8.13...
> Connected to 192.168.8.13.
> Escape character is '^]'.
> P-M0恈`Zm@!0DzaN°^C

Telnet on 9999 gives access to the admin mode, and as far as i remember i'm still on default everywhere.

> *** Channel 1
> Baudrate 4800, I/F Mode 4C, Flow 00
> Port 10001
> Connect Mode : C0
> Send '+++' in Modem Mode enabled
> Show IP addr after 'RING' enabled
> Auto increment source port disabled
> Remote IP Adr: --- none ---, Port 00000
> Disconn Mode : 00
> Flush   Mode : 00
>

**Comment by martinvw** [comment]
Looking at the code, I can understand that there is a problem, we seem to receive the first part of a message but not the rest of the message, so a read timeout feels fair.

Maybe there is a difference in behaviour between an tcp-connected rfxcom and the real rfxlan, can you enable trace and check whether it gives anything extra that can help us, otherwise I might be able to prepare a version for you with some additional logging.

**Comment by SR-G** [comment]
Yep, with trace log level i get additionnal informations, hope that will be helpful.

> 2017-09-26 17:16:17.068 [INFO ] [nternal.connector.RFXComTcpConnector] - Connecting to RFXCOM at 192.168.8.13:10001 over TCP/IP
> 2017-09-26 17:16:17.069 [DEBUG] [g.rfxcom.handler.RFXComBridgeHandler] - Reset controller
> 2017-09-26 17:16:17.069 [DEBUG] [nternal.connector.RFXComStreamReader] - Data listener started
> 2017-09-26 17:16:17.069 [TRACE] [nternal.connector.RFXComTcpConnector] - Send data (len=14): 0D00000000000000000000000000
> 2017-09-26 17:16:17.369 [DEBUG] [g.rfxcom.handler.RFXComBridgeHandler] - Get status of controller
> 2017-09-26 17:16:17.370 [TRACE] [nternal.connector.RFXComTcpConnector] - Send data (len=14): 0D00000102000000000000000000
> 2017-09-26 17:16:37.118 [ERROR] [g.rfxcom.handler.RFXComBridgeHandler] - Error occurred during packet receiving, data: 66FFFA2C6A04FFF01EFFFA2C6A04FFF0E68600FFFA2C6A04FFF060FFFA2C6A04FFF01EFFFA2C6A04FFF006FFFA2C6A04FFF07EFFFA2C6A04FFF060FFFA2C6A04FFF098F806FFFA2C6A04FFF066FFFA2C6A04FFF0E600FFFA2C6A04FFF066FFFA2C6A04FFF01EFF
> org.openhab.binding.rfxcom.internal.exceptions.RFXComUnsupportedValueException: Unsupported value '-1' for PacketType
>         at org.openhab.binding.rfxcom.internal.messages.RFXComBaseMessage$PacketType.fromByte(RFXComBaseMessage.java:95) [226:org.openhab.binding.rfxcom:2.2.0.201709261356]
>         at org.openhab.binding.rfxcom.internal.messages.RFXComMessageFactory.createMessage(RFXComMessageFactory.java:122) [226:org.openhab.binding.rfxcom:2.2.0.201709261356]
>         at org.openhab.binding.rfxcom.handler.RFXComBridgeHandler$MessageListener.packetReceived(RFXComBridgeHandler.java:233) [226:org.openhab.binding.rfxcom:2.2.0.201709261356]
>         at org.openhab.binding.rfxcom.internal.connector.RFXComBaseConnector.sendMsgToListeners(RFXComBaseConnector.java:45) [226:org.openhab.binding.rfxcom:2.2.0.201709261356]
>         at org.openhab.binding.rfxcom.internal.connector.RFXComStreamReader.run(RFXComStreamReader.java:54) [226:org.openhab.binding.rfxcom:2.2.0.201709261356]
> 2017-09-26 17:16:37.118 [ERROR] [g.rfxcom.handler.RFXComBridgeHandler] - Error occurred during packet receiving, data: 2C6A04FFF0E68600FFFA2C6A04FFF060FFFA2C6A04FFF01EFFFA2C6A04FFF006FFFA2C6A04FFF07EFFFA2C6A04
> org.openhab.binding.rfxcom.internal.exceptions.RFXComUnsupportedValueException: Unsupported value '106' for PacketType
>         at org.openhab.binding.rfxcom.internal.messages.RFXComBaseMessage$PacketType.fromByte(RFXComBaseMessage.java:95) [226:org.openhab.binding.rfxcom:2.2.0.201709261356]
>         at org.openhab.binding.rfxcom.internal.messages.RFXComMessageFactory.createMessage(RFXComMessageFactory.java:122) [226:org.openhab.binding.rfxcom:2.2.0.201709261356]
>         at org.openhab.binding.rfxcom.handler.RFXComBridgeHandler$MessageListener.packetReceived(RFXComBridgeHandler.java:233) [226:org.openhab.binding.rfxcom:2.2.0.201709261356]
>         at org.openhab.binding.rfxcom.internal.connector.RFXComBaseConnector.sendMsgToListeners(RFXComBaseConnector.java:45) [226:org.openhab.binding.rfxcom:2.2.0.201709261356]
>         at org.openhab.binding.rfxcom.internal.connector.RFXComStreamReader.run(RFXComStreamReader.java:54) [226:org.openhab.binding.rfxcom:2.2.0.201709261356]
> 2017-09-26 17:16:37.539 [DEBUG] [nternal.connector.RFXComStreamReader] - Received exception, will report it to listeners
> org.openhab.binding.rfxcom.internal.exceptions.RFXComTimeoutException: Timeout during packet read
>         at org.openhab.binding.rfxcom.internal.connector.RFXComStreamReader.processMessage(RFXComStreamReader.java:79) [226:org.openhab.binding.rfxcom:2.2.0.201709261356]
>         at org.openhab.binding.rfxcom.internal.connector.RFXComStreamReader.run(RFXComStreamReader.java:53) [226:org.openhab.binding.rfxcom:2.2.0.201709261356]
> 2017-09-26 17:16:37.540 [ERROR] [g.rfxcom.handler.RFXComBridgeHandler] - Error occurred: Timeout during packet read
> 2017-09-26 17:16:37.540 [DEBUG] [nternal.connector.RFXComStreamReader] - Data listener stopped
>

**Comment by SR-G** [comment]
(excepted that the -1 / PacketType issue has only occured once, maybe just an issue on reconnection, after that i only get : 

> 2017-09-26 17:19:17.982 [INFO ] [nternal.connector.RFXComTcpConnector] - Connecting to RFXCOM at 192.168.8.13:10001 over TCP/IP
> 2017-09-26 17:19:17.983 [DEBUG] [g.rfxcom.handler.RFXComBridgeHandler] - Reset controller
> 2017-09-26 17:19:17.983 [TRACE] [nternal.connector.RFXComTcpConnector] - Send data (len=14): 0D00000000000000000000000000
> 2017-09-26 17:19:17.983 [DEBUG] [nternal.connector.RFXComStreamReader] - Data listener started
> 2017-09-26 17:19:18.284 [DEBUG] [g.rfxcom.handler.RFXComBridgeHandler] - Get status of controller
> 2017-09-26 17:19:18.284 [TRACE] [nternal.connector.RFXComTcpConnector] - Send data (len=14): 0D00000102000000000000000000
> 2017-09-26 17:19:18.711 [DEBUG] [nternal.connector.RFXComStreamReader] - Received exception, will report it to listeners
> org.openhab.binding.rfxcom.internal.exceptions.RFXComTimeoutException: Timeout during packet read
>         at org.openhab.binding.rfxcom.internal.connector.RFXComStreamReader.processMessage(RFXComStreamReader.java:79) [226:org.openhab.binding.rfxcom:2.2.0.201709261356]
>         at org.openhab.binding.rfxcom.internal.connector.RFXComStreamReader.run(RFXComStreamReader.java:53) [226:org.openhab.binding.rfxcom:2.2.0.201709261356]
> 2017-09-26 17:19:18.711 [ERROR] [g.rfxcom.handler.RFXComBridgeHandler] - Error occurred: Timeout during packet read
> 2017-09-26 17:19:18.711 [DEBUG] [nternal.connector.RFXComStreamReader] - Data listener stopped
>

**Comment by martinvw** [comment]
Do you have the option to upgrade the firmware of your rfxlan?

**Comment by SR-G** [comment]
Sadly, no, it's an old one where the firmware upgrade could only be done by sending back the unit (which i never did).

Other than that, with OH 2.0.0-SNAPSHOT, it was roughly working at some point in time (i tryed just once but had no time to configure it at that time), items had been detected and some messages had been correctly decoded.

**Comment by mjagdis** [comment]
I believe the old RFXLans use a different protocol to the RFXCOM.
Apparently they can be reflashed with an xPL image and used with the OH1
xPL binding though. Maybe :-).

Mike

**Comment by martinvw** [comment]
Given that worked with version 2.0 it would be great to end up with a situation to support both.

@mjagdis is this something you are interested in doing?

**Pull request merged**